src/Marten/Events/Daemon/Coordination/ProjectionCoordinator.cs

```diff
@@ -0,0 +1,252 @@
+    {
+        var database = (MartenDatabase)Store.Tenancy.Default.Database;
+        if (_daemons.TryGetValue(database, out var daemon))
+        {
+            return daemon;
+        }
+
+        daemon = database.StartProjectionDaemon(Store, _logger);
+        _daemons[database] = daemon;
+
+        return daemon;
+    }
+
+    public async ValueTask<IProjectionDaemon> DaemonForDatabase(string databaseIdentifier)
+    {
+        var database = (MartenDatabase)await Store.Storage.FindOrCreateDatabase(databaseIdentifier).ConfigureAwait(false);
+        if (_daemons.TryGetValue(database, out var daemon))
+        {
+            return daemon;
+        }
+
+        daemon = database.StartProjectionDaemon(Store, _logger);
+        _daemons[database] = daemon;
+
+        return daemon;
+    }
+
+    public DocumentStore Store { get; }
+
+    public IProjectionDistributor Distributor { get; }
+
+    internal record DaemonShardName(IProjectionDaemon Daemon, ShardName Name);
+
+    public Task StartAsync(CancellationToken cancellationToken)
+    {
+        _runner = Task.Run(() => executeAsync(_cancellation.Token), _cancellation.Token);
+
+        return Task.CompletedTask;
+    }
+
+    public async Task StopAsync(CancellationToken cancellationToken)
+    {
+#if NET8_0_OR_GREATER
+        await _cancellation.CancelAsync().ConfigureAwait(false);
+#else
+        _cancellation.Cancel();
+#endif
+
+        try
+        {
+#pragma warning disable VSTHRD003
+            await _runner.ConfigureAwait(false);
+#pragma warning restore VSTHRD003
+        }
+        catch (Exception)
+        {
+            // Just trying to let it drain
+        }
+        _runner.SafeDispose();
+
+        foreach (var pair in _daemons)
+        {
+            try
+            {
+                await pair.Value.StopAllAsync().ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, "Error while trying to stop daemon agents in database {Name}", pair.Key.Identifier);
+            }
+        }
+
+        foreach (var daemon in _daemons.Values)
+        {
+            daemon.SafeDispose();
+        }
+
+        try
+        {
+            await Distributor.ReleaseAllLocks().ConfigureAwait(false);
+        }
+        catch (Exception e)
+        {
+            _logger.LogError(e, "Error trying to release subscription agent locks");
+        }
+    }
+
+    private async Task executeAsync(CancellationToken stoppingToken)
+    {
+        await Distributor.RandomWait(stoppingToken).ConfigureAwait(false);
+
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            try
+            {
+                var sets = await Distributor
+                    .BuildDistributionAsync().ConfigureAwait(false);
+
+                foreach (var set in sets)
+                {
+                    // Is it already running here?
+                    if (Distributor.HasLock(set))
```

Comment:
**CONCERN:** Do we need this if on HasLock? I think that just calling `TryAttainLock` should be enough; if we don't get it, then we're good.

Reply:
I wasn't sure that the call through to try to attain the lock would be repeatable to be honest. This is specifically for the case of "we own this projection (shard), but it's been paused, so try to restart it"